graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -109,17 +110,21 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
                     case REGEX:
                         regexRules.add(rule);
                         break;
+                    case CONTAINS:
+                        containsRules.add(rule);
+                        break;
                 }
             }
         }
 
-        final int size = presenceRules.size() + exactRules.size() + greaterRules.size() + smallerRules.size() + regexRules.size();
+        final int size = presenceRules.size() + exactRules.size() + greaterRules.size() + smallerRules.size() + regexRules.size() + containsRules.size();
         this.rulesList = Lists.newArrayListWithCapacity(size);
         this.rulesList.addAll(presenceRules);
         this.rulesList.addAll(exactRules);
         this.rulesList.addAll(greaterRules);
         this.rulesList.addAll(smallerRules);
         this.rulesList.addAll(regexRules);
+        this.rulesList.addAll(containsRules);
```

Comment:
The order of the list additions are ordered by their complexity, as this allows us to skip some rule evaluations in some cases (preferably the heavy-hitters). Due to this, the `containsRules` should be added earlier, at least before the `regexRules`.